clients/src/main/java/org/apache/kafka/clients/consumer/StickyAssignor.java

```diff
@@ -0,0 +1,591 @@
+                for (TopicPartition topicPartition: entry.getValue())
+                    currentPartitionConsumer.remove(topicPartition);
+                it.remove();
+            } else {
+                // otherwise (the consumer still exists)
+                for (Iterator<TopicPartition> partitionIter = entry.getValue().iterator(); partitionIter.hasNext();) {
+                    TopicPartition partition = partitionIter.next();
+                    if (!partition2AllPotentialConsumers.containsKey(partition)) {
+                        // if this topic partition of this consumer no longer exists remove it from currentAssignment of the consumer
+                        partitionIter.remove();
+                        currentPartitionConsumer.remove(partition);
+                    } else
+                        // otherwise, remove the topic partition from those that need to be assigned (because it is already assigned
+                        // and we would want to preserve that assignment as much as possible)
+                        unassignedPartitions.remove(partition);
+                }
+            }
+        }
+        // at this point we have preserved all valid topic partition to consumer assignments and removed
+        // all invalid topic partitions and invalid consumers. Now we need to assign unassignedPartitions
+        // to consumers so that the topic partition assignments are as balanced as possible.
+
+        // an ascending sorted set of consumers based on how many topic partitions are already assigned to them
+        TreeSet<String> sortedCurrentSubscriptions = new TreeSet<>(new SubscriptionComparator(currentAssignment));
+        sortedCurrentSubscriptions.addAll(currentAssignment.keySet());
+
+        balance(sortedPartitions, unassignedPartitions, sortedCurrentSubscriptions, consumer2AllPotentialPartitions,
+                partition2AllPotentialConsumers, oldAssignment, currentPartitionConsumer);
+        return currentAssignment;
+    }
+
+
+    @Override
+    protected ByteBuffer getUserData() {
+        try {
+            ByteArrayOutputStream bos = new ByteArrayOutputStream(2048);
+            ObjectOutputStream oos = new ObjectOutputStream(bos);
+            oos.writeObject(currentAssignment);
+            return ByteBuffer.wrap(bos.toByteArray());
+        } catch (Exception ioex) {
+            log.error("Failed to serialize currentAssignment", ioex);
+            return super.getUserData();
+        }
+    }
+
+    @SuppressWarnings("unchecked")
+    @Override
+    public void onAssignment(Assignment assignment) {
+        try {
+            byte[] array = new byte[assignment.userData().remaining()];
+            assignment.userData().get(array);
+            ObjectInputStream ois = new ObjectInputStream(new ByteArrayInputStream(array));
+            currentAssignment = (Map<String, List<TopicPartition>>) ois.readObject();
```

Comment:
Unless I'm misunderstanding something, it seems like we're giving the full group assignment to every member in the group. I expected instead that each member would only receive its own assignment for the current generation and that we would aggregate the individual assignments on the leader when we received the group subscriptions. If we send all the assignments, then the overall overhead grows quadratically with the number of members in the group.